web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx

```diff
@@ -0,0 +1,280 @@
+
+import SearchPanel from 'teleport/UnifiedResources/SearchPanel';
+
+import { UnifiedResourceResponse } from 'teleterm/services/tshd/types';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import * as uri from 'teleterm/ui/uri';
+import { useWorkspaceContext } from 'teleterm/ui/Documents';
+import { useWorkspaceLoggedInUser } from 'teleterm/ui/hooks/useLoggedInUser';
+import { useConnectMyComputerContext } from 'teleterm/ui/ConnectMyComputer';
+
+import { retryWithRelogin } from 'teleterm/ui/utils';
+
+import {
+  ConnectServerActionButton,
+  ConnectKubeActionButton,
+  ConnectDatabaseActionButton,
+} from './actionButtons';
+import { useResourcesContext } from './resourcesContext';
+
+interface UnifiedResourcesProps {
+  clusterUri: uri.ClusterUri;
+}
+
+export function UnifiedResources(props: UnifiedResourcesProps) {
+  const appContext = useAppContext();
+  const { onResourcesRefreshRequest } = useResourcesContext();
+
+  const [params, setParams] = useState<UnifiedResourcesQueryParams>({
+    sort: { fieldName: 'name', dir: 'ASC' },
+  });
+
+  const { documentsService, rootClusterUri } = useWorkspaceContext();
+  const loggedInUser = useWorkspaceLoggedInUser();
+  const { canUse: hasPermissionsForConnectMyComputer, agentCompatibility } =
+    useConnectMyComputerContext();
+
+  const isRootCluster = props.clusterUri === rootClusterUri;
+  const canAddResources = isRootCluster && loggedInUser?.acl?.tokens.create;
+
+  const canUseConnectMyComputer =
+    isRootCluster &&
+    hasPermissionsForConnectMyComputer &&
+    agentCompatibility === 'compatible';
+
+  const { fetch, resources, attempt, clear } = useUnifiedResourcesFetch({
+    fetchFunc: useCallback(
+      async (paginationParams, signal) => {
+        const response = await retryWithRelogin(
+          appContext,
+          props.clusterUri,
+          () =>
+            appContext.resourcesService.listUnifiedResources(
+              {
+                clusterUri: props.clusterUri,
+                searchAsRoles: false,
+                sortBy: {
+                  isDesc: params.sort.dir === 'DESC',
+                  field: params.sort.fieldName,
+                },
+                search: params.search,
+                kindsList: params.kinds,
+                query: params.query,
+                pinnedOnly: params.pinnedOnly,
+                startKey: paginationParams.startKey,
+                limit: paginationParams.limit,
+              },
+              signal
+            )
+        );
+
+        return {
+          startKey: response.nextKey,
+          agents: response.resources,
+          totalCount: response.resources.length,
+        };
+      },
+      [appContext, params, props.clusterUri]
+    ),
+  });
+
+  useEffect(() => {
+    const { cleanup } = onResourcesRefreshRequest(() => {
+      clear();
+      fetch({ force: true });
+    });
+    return cleanup;
+  }, [onResourcesRefreshRequest, fetch, clear]);
```

Comment:
I think refreshing the resources might be not working as expected. After I set up Connect My Computer, the unified view is refreshed only after I open the tab, not in the background.

(Ignore the error when removing the token, I purposefully disabled that on my cluster a while ago but it works the same even if I can remove join tokens)


https://github.com/gravitational/teleport/assets/27113/0af4b7d9-0b13-4b88-82a9-057740852ef9

Reply:
I'm able to reproduce it.
It happens because `fetchInternal` in `useKeyBasedPagination` checks `finished` state. If you previously opened resources tab and loaded all results this flag is `true`, so the fetch function aborts early. We added `clear` function to reset all state, but unfortunately `fetch` still sees the old state. I think it can be a result of two things:
* We pass `fetch` to `useEffect` when registering the callback. After `clear` is done, we call `fetch`, but it is the 'old' fetch.
* `setState` is async. We call `fetch` immediately after `clear`, I guess the `finished` state change may not be 'propagated' yet. 

When I was refactoring `useKeyBasedPagination` I made an attempt to refactor the `fetch` function, because it seemed to me that it might have bugs (in short, it calculates the new state based on the old state, but it doesn't use update functions for it; we should rather keep most of its state in `useRef`). In the end I gave up because it was quite a lot of changes and I was afraid of breaking something :/ 
Now I will have to go back to it, I will add a task in a tracking issue. Thanks a lot for catching it!